odl/trafos/wavelet.py

```diff
@@ -834,14 +957,511 @@ def adjoint(self):
         if self.is_orthogonal:
             return self.inverse
         else:
-            # TODO: put adjoint here
-            return super().adjoint
+            if not isinstance(self.wbasis, pywt.Wavelet):
+                return InverseAdjBiorthWaveletTransform(dom=self.range,
+                                                        nscales=self.nscales,
+                                                        wbasis=self.wbasis)
+            else:
+                return super().adjoint
 
     @property
     def inverse(self):
         """The inverse wavelet transform."""
-        return WaveletTransform(dom=self.range, nscales=self.nscales,
-                                wbasis=self.wbasis, mode=self.mode)
+        if not isinstance(self.wbasis, pywt.Wavelet):
+            return BiorthWaveletTransform(dom=self.range, ran=self.domain,
+                                          nscales=self.nscales,
+                                          wbasis=self.wbasis)
+        elif isinstance(self.wbasis, pywt.Wavelet):
+            return WaveletTransform(dom=self.range, nscales=self.nscales,
+                                    wbasis=self.wbasis, mode=self.mode)
+
+
+class BiorthWaveletTransform(Operator):
```

Comment:
I think the main line of thought here still should be to move all of this JOS things outside of ODL and into a fully separate operator, then it could be called `BiorthWaveletTransform` without problems.

If we wanted to, we could make an abstract "WaveletTransform" class that both of these inherit from (we could even overload its `__new__` to return a proper one depending on inputs).